_posts/2024-08-01-search-standalone-mapper.adoc

```diff
@@ -0,0 +1,295 @@
+we will rely on search projections instead. And projections do not need entity loading.
+Now, if the identifier loader provides actual search entities, the entity loader has nothing more to do
+but just pass through the batch of received "identifiers", which are actual entities, to the entity sink.
+
+With that in mind, the mass-loading strategy may be implemented as:
+
+[source, java]
+====
+----
+new MassLoadingStrategy<Guide, Guide>() {
+    @Override
+    public MassIdentifierLoader createIdentifierLoader(LoadingTypeGroup<Guide> includedTypes,
+            MassIdentifierSink<Guide> sink, MassLoadingOptions options) {
+            // ...  // <1>
+        };
+    }
+
+    @Override
+    public MassEntityLoader<Guide> createEntityLoader(LoadingTypeGroup<Guide> includedTypes,
+            MassEntitySink<Guide> sink,
+            MassLoadingOptions options) {
+        return new MassEntityLoader<Guide>() { // <2>
+            @Override
+            public void close() {
+                // noting to do
+            }
+
+            @Override
+            public void load(List<Guide> guides) throws InterruptedException {
+                sink.accept(guides); // <3>
+            }
+        };
+    }
+})
+----
+1. We'll look at the implementation of the identifier loader in the following code snippet as
+it is slightly trickier than the pass-through entity loader.
+Hence, we would want to take a closer look at it.
+2. An implementation of the pass-through entity loader.
+3. As explained above, we treat the search entities as identifiers and simply pass the entities we receive to the sink.
+====
+
+Since the guide search entities are constructed from the information read from a file,
+we have to somehow pass the stream of these guides to the identifier loader.
+We could do this by using the `MassLoadingOptions options`.
+These mass loading options provide access to the context objects passed to the mass indexer by the user.
+
+[source, java]
+====
+----
+@Inject
+SearchMapping searchMapping; // <1>
+// ...
+
+var future = searchMapping.scope(Object.class) // <2>
+    .context(GuideLoadingContext.class, guideLoadingContext) // <3>
+    // ... <4>
+    .massIndexer() // <5>
+    .start(); // <6>
+----
+1. Inject `SearchMapping` somewhere in your application.
+2. Create a scope, as usual.
+3. Pass the context object to the mass indexer that knows how to read the files,
+and is able to provide the batches of `Guide` search entities. See the following code snippet
+for an example of how such context can be implemented.
+4. Set any other mass indexer configuration options as needed.
+5. Create a mass indexer.
+6. Start the indexing process.
+----
+public class GuideLoadingContext {
+
+    private final Iterator<Guide> guides;
+
+    GuideLoadingContext(Stream<Guide> guides) {
+        this.guides = guides.iterator(); // <1>
+    }
+
+    public synchronized List<Guide> nextBatch(int batchSize) {
```

Comment:
Is `synchronized` really necessary here? There's only one thread producing identifiers, IIRC.

Reply:
+1 yeah, only one thread loading the ids... 

https://github.com/hibernate/hibernate-search/blob/2126a7598159ca8a9fa59e2db18d718bed5cc13b/mapper/pojo-base/src/main/java/org/hibernate/search/mapper/pojo/massindexing/impl/PojoMassIndexingBatchIndexingWorkspace.java#L109-L113

I'll open a PR to make an update in search,quarkus.io ...